server/src/main/java/org/opensearch/cluster/store/RemoteClusterStateService.java

```diff
@@ -0,0 +1,185 @@
+                    .add("cluster-state")
+                    .add(clusterUUID)
+                    .add("index")
+                    .add(indexUUID)
+            );
+    }
+
+    private BlobContainer markerContainer(String clusterName, String clusterUUID) {
+        // 123456789012_test-cluster/cluster-state/dsgYj10Nkso7/marker
+        initializeRepository();
+        return blobStoreRepository.blobStore()
+            .blobContainer(
+                blobStoreRepository.basePath()
+                    .add(Base64.getUrlEncoder().withoutPadding().encodeToString(clusterName.getBytes(StandardCharsets.UTF_8)))
+                    .add("cluster-state")
+                    .add(clusterUUID)
+                    .add("marker")
+            );
+    }
+
+    // All the changes above this point are from another PR - https://github.com/opensearch-project/OpenSearch/pull/9160
+    // TODO - Remove above comment after changes in above mentioned PR are merged to main
+
+    /**
+     * Fetch latest index metadata which is part of remote cluster state
+     * @param clusterUUID uuid of cluster state to refer to in remote
+     * @param clusterName name of the cluster
+     * @return latest IndexMetadata map
+     */
+    public Map<String, IndexMetadata> getLatestIndexMetadata(String clusterUUID, String clusterName) throws IOException {
+        Map<String, IndexMetadata> remoteIndexMetadata = new HashMap<>();
+        ClusterMetadataMarker clusterMetadataMarker = getLatestClusterMetadataMarker(clusterUUID, clusterName);
+        for (Map.Entry<String, UploadedIndexMetadata> entry : clusterMetadataMarker.getIndices().entrySet()) {
+            IndexMetadata indexMetadata = INDEX_METADATA_FORMAT.read(
+                indexMetadataContainer(clusterName, clusterUUID, entry.getValue().getIndexUUID()),
+                entry.getValue().getUploadedFilename(),
+                blobStoreRepository.getNamedXContentRegistry()
+            );
+            remoteIndexMetadata.put(entry.getKey(), indexMetadata);
+        }
+        return remoteIndexMetadata;
+    }
+
+    private ClusterMetadataMarker getLatestClusterMetadataMarker(String clusterUUID, String clusterName) {
+        String latestMarkerFileName = getLatestMarkerFileName(clusterUUID, clusterName);
+        return fetchRemoteClusterMetadataMarker(latestMarkerFileName, clusterUUID, clusterName);
+    }
+
+    private String getLatestMarkerFileName(String clusterUUID, String clusterName) throws IllegalStateException {
+        try {
+            List<BlobMetadata> markerFilesMetadata = markerContainer(clusterUUID, clusterName).listBlobsByPrefixInSortedOrder(
+                "marker",
+                1,
+                BlobContainer.BlobNameSortOrder.LEXICOGRAPHIC
+            );
+            if (markerFilesMetadata != null && !markerFilesMetadata.isEmpty()) {
+                return markerFilesMetadata.get(0).name();
+            }
+        } catch (IOException e) {
```

Comment:
I think its FileNotFoundException, but let me confirm. Do we need to handle this case differently or change the error logging?

Reply:
the exception depends on the type of repository, it could be FileNotFoundException, NoSuchFileException. But the exception would be a sub class of IOException.